File: Editorial/99_Recover_Binary_Search_Tree/Go/3_Recursive_Inorder_Traversal.go
func recoverTree(root *TreeNode) {
    var x, y, pred *TreeNode
    findTwoSwapped(root, &x, &y, &pred)
    // Swap x and y values
    x.Val, y.Val = y.Val, x.Val
}

func findTwoSwapped(root *TreeNode, x, y, pred **TreeNode) {
    if root == nil {
        return
    }
    findTwoSwapped(root.Left, x, y, pred)
    if *pred != nil && root.Val < (*pred).Val {
        *y = root
        if *x == nil {
            *x = *pred
        } else {
            return
        }
    }
    *pred = root
    findTwoSwapped(root.Right, x, y, pred)
}

File: Editorial/99_Recover_Binary_Search_Tree/TypeScript/3_Recursive_Inorder_Traversal.ts
function recoverTree(root: TreeNode | null): void {
    // first node
    let x: TreeNode | null = null;
    // second node
    let y: TreeNode | null = null;
    // predecessor node
    let pred: TreeNode | null = null;
    function findTwoSwapped(root: TreeNode | null): void {
        if (root === null) return;
        findTwoSwapped(root.left);
        if (pred !== null && root.val < pred.val) {
            y = root;
            if (x === null) x = pred;
            else return;
        }
        pred = root;
        findTwoSwapped(root.right);
    }
    findTwoSwapped(root);
    // Swap x and y values
    [x.val, y.val] = [y.val, x.val];
}
